### PR TITLE
feat: replace Jaeger exporter with OTLP

### PR DESCRIPTION
**Pull request by Rauno56**
# Description

I'm making an assumption that Smart Agent does not support OTLP. Am I right on that?

- APMI-1932

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Tested manually
- [x] Added automated tests

# Checklist:

- [x] Changelogs have been updated
- [x] Unit tests have been added/updated
- [x] Documentation has been updated

### CODE REVIEW

**Comment by Rauno56** [comment]
Converted it back to draft since Synthetic Run tests started failing. Probably has to do with updated deps and not the changed exporter.

File: MIGRATING.md
Comment (owais) [inline MIGRATING.md:77]: This is not a 1-1 mapping. We should add comment that this needs to be OTLP endpoint OR they need to use `OTEL_EXPORTER_JAEGER_ENDPOINT`

**Comment by owais** [comment]
https://github.com/signalfx/gdi-specification/blob/main/specification/configuration.md#opentelemetry-environment-variable 

We need to support `OTEL_TRACES_EXPORTER`. Check GDI spec for details ^

**Comment by Rauno56** [comment]
> Converted it back to draft since Synthetic Run tests started failing. Probably has to do with updated deps and not the changed exporter.

It was a faulty test: https://github.com/signalfx/splunk-otel-js/pull/228

**Comment by Rauno56** [comment]
> [https://github.com/signalfx/gdi-specification/blob/main/specification/configuration.md#opentelemetry-environment-variable](https://github.com/signalfx/gdi-specification/blob/main/specification/configuration.md?rgh-link-date=2021-08-09T11%3A23%3A22Z#opentelemetry-environment-variable)
> 
> We need to support `OTEL_TRACES_EXPORTER`. Check GDI spec for details ^

Thanks for pointing that out! I see that something that needs to be added on top of the default, which this PR changes. Do you think we could add it in another PR?

File: MIGRATING.md
Comment (Rauno56) [reply MIGRATING.md:77]: True, but aren't we asusming they're migrating to Splunk APM which, in a way, is an equivalent with OTLP-support?

Comment (Rauno56) [reply MIGRATING.md]: I added https://github.com/signalfx/splunk-otel-js/pull/218/commits/7d3db551434cc4c063455d46494f1e01926b2700 for now, but rather remove it regarding my comment above.

**Comment by owais** [comment]
> Thanks for pointing that out! I see that something that needs to be added on top of the default, which this PR changes. Do you think we could add it in another PR?

We could but it's better if such changes land together so we can review everything at once. May be land that feature first and then merge this PR? That might make it easier. 

Either way, totally up to you.

**Comment by Rauno56** [comment]
> > Thanks for pointing that out! I see that something that needs to be added on top of the default, which this PR changes. Do you think we could add it in another PR?
> 
> We could but it's better if such changes land together so we can review everything at once. May be land that feature first and then merge this PR? That might make it easier.
> 
> Either way, totally up to you.

Since the feature of controlling the exporter via env var was not there before this PR, I'd merge this first and then make it easily configurable later. It makes more sense to me because the changes of this PR will need to be merged as is either way and the support for `OTEL_TRACES_EXPORTER` is additive to this - otherwise I'd need to possibly significantly change this PR later.

Please approve, if that sounds reasonable.

**Comment by owais** [comment]
OK but you'll need to ensure no one else releases this before your next PR lands :)

File: src/tracing.ts
Comment (owais) [inline src/tracing.ts:40]: why are we removing this?

File: examples/express/.env.example
Comment (owais) [inline examples/express/.env.example:3]: Do we use OTLP HTTP or gRPC exporter now and is it supported by SFX ingest?

**Review by owais — CHANGES_REQUESTED** [review]
I took another look and it seems we are entirely dropping support for jaeger. That doesn't look right to me. I think this PR should just add support for OTLP exporter without removing any support for Jaeger.

File: examples/express/.env.example
Comment (Rauno56) [reply examples/express/.env.example:3]: We are using OTLP/HTTP which is supported by the endpoint listed :point_up:

**Comment by Rauno56** [comment]
@owais, Jaeger support is now back behind `OTEL_TRACES_EXPORTER=jaeger-thrift-splunk` env var or 
`tracesExporter` config option.

File: examples/express/utils.js
Comment (owais) [inline examples/express/utils.js:25]: I don't think we support SIGNALFX_ENDPOINT_URL in new Otel based libs anymore.

Comment (owais) [reply examples/express/utils.js:25]: I think we should remove these as the libs don't support these fallbacks and it might be confusing to keep them in examples.

File: src/options.ts
Comment (owais) [inline src/options.ts:49]: should this be jaeger-thrift or jaeger_thrift?

File: src/options.ts
Comment (owais) [inline src/options.ts:51]: do we want this to be `otlp-grpc`?

File: src/options.ts
Comment (owais) [inline src/options.ts:62]: we already have a spanExporterFactory option below, do we really need this?

File: src/options.ts
Comment (owais) [inline src/options.ts:128]: options are passed as an argument to spanExporterFactory. Each factory should host code to parse/set defaults internally.

File: src/options.ts
Comment (owais) [inline src/options.ts:123]: I think this should be purely an env var, not an argument to the function as we already have spanExporterFactory argument.

File: src/options.ts
Comment (owais) [reply src/options.ts:128]: To clarify, I mean this function should not have specific knowledge about each exporter. it should just pass on the options instance to the exporter factory. Then exporter factory should read it's relevant env vars, options.endpoint, etc.

File: src/options.ts
Comment (owais) [inline src/options.ts:129]: I think jaeger and jaeger-thrift-splunk should have two different exporter factories each with their own default endpoints.

Comment (owais) [reply src/options.ts:129]: Or may be a single factory that accepts the default endpoint as an argument.

File: src/options.ts
Comment (owais) [inline src/options.ts:165]: I think we should call this `otlpSpanExporterFactory` instead of default. The constructor decides what is default and what isn't. Factory itself has no business knowing if it is used as the default one.

**Review by owais — CHANGES_REQUESTED** [review]
Overall direction looks good. Some superficial comments about public API, naming, etc.

File: examples/express/utils.js
Comment (Rauno56) [reply examples/express/utils.js:25]: This whole function is specifically for people migrating as the first preliminary step to be able to..
1. get something that looks like their old config up and running with OTEL;
2. see what needs to be converted and how.

File: src/options.ts
Comment (Rauno56) [reply src/options.ts:51]: It's HTTP, ingress doesn't currently support grpc.
[The spec](https://github.com/signalfx/gdi-specification/blob/aea51ebe986de6980950ca3f7948d84420960d8b/specification/configuration.md) says that the default has to be `otlp`.

Comment (owais) [reply src/options.ts:51]: Sorry I didn't mean gpprc specifically but just to add the format instead of just calling it otlp but if the spec recommends otlp then I suppose we use otlp as an alias for otlp-http.

File: examples/express/utils.js
Comment (owais) [reply examples/express/utils.js:25]: I don't think we ever had a goal to have config backward compatibility in Otel for old SFX libs. IMO this can lead to bugs as things will work with SIGNALFX config in this config but may not in another real project. IMO we should restrict the migration guide strictly to markdown documentation and not pollute otel examples with old sfx config. I'm sure this will lead to confusion. 

That said, I see you're just updating existing examples and didn't add new one so it's not a blocker for me for this PR. I still think we should remove references to SFX stuff later and keep pure otel examples here.

Comment (Rauno56) [reply examples/express/utils.js:25]: > I don't think we ever had a goal to have config backward compatibility in Otel for old SFX libs.

Not backwards compatibility, but helping through testing and conversion. I'd rather get some feedback from the first users, but totally open to removing this if it ever causes any confusion.

File: src/options.ts
Comment (Rauno56) [reply src/options.ts:123]: Resolving this because it's the same argument as https://github.com/signalfx/splunk-otel-js/pull/218#discussion_r690266320

File: src/options.ts
Comment (Rauno56) [reply src/options.ts:128]: Rest of `_setDefaultOptions` sets and parses defaults for other parts of the API and we already pass the factories the whole `options` object instead of letting it take some of it's configuration from `options` and rest from the env. I don't see how exporters are different.

File: src/options.ts
Comment (Rauno56) [reply src/options.ts:49]: Since this is not in the spec, I'll make that more specific as you suggest using the key from OTel collector: `jaeger-thrift-http`

https://github.com/open-telemetry/opentelemetry-collector/blob/main/examples/k8s/otel-config.yaml

File: src/options.ts
Comment (Rauno56) [reply src/options.ts:51]: I get where you are coming from, but I'd not add aliasing yet and leave only the minimal implementation per spec for now.

File: src/options.ts
Comment (Rauno56) [reply src/options.ts:62]: We could live without it, but I think it's useful to have as much parity between different formats of configuration as possible. Sure, we cannot let people set functions through env vars which is why I'm open to removing this if you'd really prefer that and have it conceptually be a alternative to setting the function in "`config.js`" BUT through environment. WDYT?

File: src/options.ts
Comment (Rauno56) [reply src/options.ts:129]: See https://github.com/signalfx/splunk-otel-js/pull/218#discussion_r693729544

File: src/options.ts
Comment (owais) [reply src/options.ts:62]: I feel this is cluttering the constructor unnecessarily. Now we are providing two different ways of configuring an exporter with code. What if both are defined? Which one takes precedence? How do users decide which one to use when? If we accept tracesExporter as a well-known string, why not processor as well? What happens when we add metrics support? Do we add two options per component again?  This raises too many questions and doesn't solve enough problems (IMO) to be worth it. I think we should remove it and continue to use factory pattern.

File: src/options.ts
Comment (owais) [reply src/options.ts:128]: What I meant was that the logic for configuring each exporter is leaking outside the factory to the constructor. I think constructor should do as little as possible aand only try to identify the components to be used. Then let the components configure themselves, including handling their own defaults, validation logic etc.

File: src/options.ts
Comment (Rauno56) [reply src/options.ts]: Fair. Removed it. In doing so, introducing `otlp-http` alias became trivial, so I did that as well.

File: src/options.ts
Comment (Rauno56) [reply src/options.ts]: See https://github.com/signalfx/splunk-otel-js/pull/218#discussion_r695641719

File: src/options.ts
Comment (Rauno56) [reply src/options.ts]: Split the factories up.

File: src/options.ts
Comment (Rauno56) [reply src/options.ts]: Changed that. Thanks.

File: src/options.ts
Comment (owais) [inline src/options.ts:148]: nit: IMO the following would be much simpler to read for most people including me but up to you if you want to change

```javascript

const factory = spanExporterMap[process.env.OTEL_TRACES_EXPRTER ?? 'default']
if  (factory == undefined) {
    throw new Error(...)
}
return factory
```

**Review by owais — COMMENTED** [review]
Looks good. Left a couple of minor comments

File: src/options.ts
Comment (Rauno56) [reply src/options.ts:148]: If you refer to the usage of `assert` then `if (...) { throw new Error(...); }` is exactly what that does - I do agree that it's somewhat less common outside of tests than if-error, but I do prefer that over if-errors if that's fine by you.
Or do you mean that the error message is long-winding?

File: src/options.ts
Comment (Rauno56) [reply src/options.ts]: I did not mind the duplication, but don't have a strong opinion on that. Changed it per your comment.

File: src/options.ts
Comment (owais) [reply src/options.ts:148]: I don't mind the assert per se but I feel simple `if` condition to be much easier read and understand than a function that takes callbacks, type of object, a literal "function", etc. Anyone not an expert in JS/TS will have to spent a few minutes trying to understand it and probably read docs on assert function. On the other hand, if condition is universal and needs 5 second to understand even if you've never written JS before.

File: examples/express/utils.js
Comment (owais) [reply examples/express/utils.js]: I think we should have separate migration example and keep this one as pure Otel based but dont need it in this PR.

**Review by owais — APPROVED** [review]

**Comment by owais** [comment]
@jtmalinowski please take a look as well. Thanks

File: src/options.ts
Comment (Rauno56) [reply src/options.ts]: We do not. Removed the export. Thanks!

**Comment by Rauno56** [comment]
@jtmalinowski gave his approval offline.

**Pull request merged**
